runtime-sdk/modules/evm/src/raw_tx.rs

```diff
@@ -0,0 +1,411 @@
+use std::convert::TryInto;
+
+use anyhow::{anyhow, Context as _};
+use ethereum;
+use k256;
+use sha3::{self, Digest as _};
+
+use oasis_runtime_sdk::{
+    crypto::signature,
+    types::{token, transaction},
+};
+
+use crate::types;
+
+pub fn recover_low(
+    sig: &k256::ecdsa::recoverable::Signature,
+    sig_digest: sha3::Keccak256,
+) -> Result<k256::ecdsa::VerifyingKey, anyhow::Error> {
+    if sig.s().is_high().into() {
+        return Err(anyhow!("signature s high"));
+    }
+    sig.recover_verify_key_from_digest(sig_digest)
+        .with_context(|| "recover verify key from digest")
+}
+
+fn decode_enveloped(
+    body: &[u8],
+    expected_chain_id: Option<u64>,
+) -> Result<transaction::Transaction, anyhow::Error> {
+    let (
+        chain_id,
+        sig,
+        sig_digest,
+        eth_action,
+        eth_value,
+        eth_input,
+        eth_nonce,
+        eth_gas_price,
+        eth_gas_limit,
+    ) = match rlp::decode::<ethereum::TransactionV2>(body)
+        .with_context(|| "decoding transaction rlp")?
+    {
+        ethereum::TransactionV2::Legacy(eth_tx) => {
+            let sig = k256::ecdsa::recoverable::Signature::new(
+                &k256::ecdsa::Signature::from_scalars(
+                    eth_tx.signature.r().to_fixed_bytes(),
+                    eth_tx.signature.s().to_fixed_bytes(),
+                )
+                .with_context(|| "signature from_scalars")?,
+                k256::ecdsa::recoverable::Id::new(eth_tx.signature.standard_v())
+                    .with_context(|| "recoverable id new")?,
+            )
+            .with_context(|| "recoverable signature new")?;
+            let message = ethereum::LegacyTransactionMessage::from(eth_tx);
+
+            // TODO: replace this with message.hash()
+            let mut sig_digest = sha3::Keccak256::new();
+            sig_digest.update(&rlp::encode(&message));
+
+            (
+                message.chain_id,
```

Comment:
Oh so at some point there was no domain separation? We should just reject these cases without chain ID.

Reply:
can't, we use a bunch of sigs without chain id in the test suite

Reply:
Sounds good.